### PR TITLE
Return LB picks with an error upon shutdown.

### LINKED ISSUE #12506: Call dropped by load balancing policy
### What version of gRPC and what language are you using?
gRPC 1.6.0 installed by pip
### What operating system (Linux, Windows, …) and version?
CentOS Linux release 7.3.1611 (Core)
### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.4.5 (default, Nov  9 2016, 16:24:59)
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
 
### What did you do?
I have a series of py.test suites that check GRPC API of a particular service. Channel and client for a service is periodically recreated in a such manner:
```python
class DummyGRPCChannel(object):
    def __init__(self, host, port, root_certificates=None, private_key=None, certificate_chain=None, logger=logging):
        self.host = host
        self.port = port
        self.endpoint = join_host_port(self.host, self.port)
        self.credentials = grpc.ssl_channel_credentials(root_certificates, private_key, certificate_chain)
        self.channel = self._connect(logger=logger)
        self.stubs = dict()

    def _create_channel(self):
        raise Exception('Not implemented')

    def _connect(self, logger=logging):
        logger.info('Trying to connect to {}'.format(self.endpoint))
        with step('[PRECONDITION] Waiting for grpc channel at {}'.format(self.endpoint)):
            wait_for_connection(host=self.host, port=self.port)
        return self._create_channel()

    def create_stub(self, stub_cls):
        _stub = stub_cls(self.channel)
        _stub_name = stub_cls.__name__
        if _stub_name not in self.stubs:
            self.stubs[_stub_name] = list()
        self.stubs[_stub_name].append(_stub)
        return _stub

class GRPCChannel(DummyGRPCChannel):
    def _create_channel(self):
        channel = grpc.insecure_channel(
            self.endpoint,
            options=[('grpc.max_message_length', Constants.MAX_MESSAGE_LEN), ('grpc.min_reconnect_backoff_ms', 100)])
        return channel

class DummyEndpoint(object):
    def __init__(self, host, port, use_tls=False, tls_creds=None):
        self.host = host
        self.port = port
        self._channel = None
        self.clients = dict()
        self.use_tls = use_tls
        self.tls_creds = tls_creds

class Endpoint(DummyEndpoint):
    def _connect(self, logger=logging):
        if self.use_tls:
            self._channel = SSLGRPCChannel(self.host, self.port, logger=logger, **self.tls_creds)
        else:
            self._channel = GRPCChannel(self.host, self.port, logger=logger)
    # Method for create client
    def create_client(self, stub, name=None, logger=logging):
        if self._channel is None:
            self._connect(logger=logger)
        name = name if name is not None else rand_str()
        with step('Creating {} client'.format(stub.__name__)):
            _client = self._channel.create_stub(stub)
            self.clients[name] = _client
            return _client
```
Using:
```python
from component import component_pb2

service = Endpoint(HOST, PORT, False)
client = service.create_client(component_pb2.ComponentStub, name='component', logger=logger)
client.DoSomething(component_pb2.DoSomethingRequest())
```
Requests to the service are performed in a default manner (according to the grpcio documentation). There are a lot of tests within a single testing session, and after a while, requests start to fail with a strange error: "Call dropped by load balancing policy". This problem is floating, so the moment when error start to occur is changes from time to time.
 
### What did you expect to see?
*Successful response from the server*
 
### What did you see instead?
```python
state = <grpc._channel._RPCState object at 0x7fd982bfbda0>, call = <grpc._cython.cygrpc.Call object at 0x7fd982bfb9e8>, with_call = False, deadline = None
    def _end_unary_response_blocking(state, call, with_call, deadline):
        if state.code is grpc.StatusCode.OK:
            if with_call:
                rendezvous = _Rendezvous(state, call, None, deadline)
                return state.response, rendezvous
            else:
                return state.response
        else:
>           raise _Rendezvous(state, None, None, deadline)
E           grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNKNOWN, Call dropped by load balancing policy)>
/usr/lib64/python3.4/site-packages/grpc/_channel.py:440: _Rendezvous
```
The thing that confuses me a lot is that every ```GRPC_ERROR_NONE``` is treated as load balancing error in GRPC core:
```
    calld->error = error == GRPC_ERROR_NONE
                       ? GRPC_ERROR_CREATE_FROM_STATIC_STRING(
                             "Call dropped by load balancing policy")
                       : GRPC_ERROR_CREATE_REFERENCING_FROM_STATIC_STRING(
                             "Failed to create subchannel", &error, 1);
```
See https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.c#L1027

### PR DESCRIPTION
**Pull request by markdroth**
Fixes the error message returned in the case described in #12506.  (Does not actually fix the underlying bug.)

Returning a pick with a target of NULL but no error is interpreted by the client_channel code as the LB policy dropping the call:

https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.cc#L1033

We should not be doing that in any other case.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +248 [None]                                                                                  +648  +0.0%
  +1.0%    +144 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc                          +144  +1.0%
       +36%    +137 glb_shutdown_locked                                                                     +137   +36%
      +1.9%      +7 [Unmapped]                                                                                +7  +1.9%
  +2.1%    +128 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc                  +128  +2.1%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, pick_first_lb_policy*) [clone .isra.    +192  [NEW]
  +0.4%     +32 src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc                 +32  +0.4%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, round_robin_lb_policy*) [clone .isra    +192  [NEW]
      +6.8%      +9 [Unmapped]                                                                                +9  +6.8%

  +0.0%    +552 TOTAL                                                                                   +952  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 1]
```diff
@@ -1033,15 +1033,17 @@ static void glb_shutdown_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol) {
...
     pp = next;
   }
 
   while (pping != NULL) {
     pending_ping *next = pping->next;
-    GRPC_CLOSURE_SCHED(exec_ctx, &pping->wrapped_notify_arg.wrapper_closure,
-                       GRPC_ERROR_NONE);
+    GRPC_CLOSURE_SCHED(
+        exec_ctx, &pping->wrapped_notify_arg.wrapper_closure,
+        GRPC_ERROR_CREATE_FROM_STATIC_STRING("Channel Shutdown"));
```
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: We may also want to add `gpr_free(pping);` here.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc [hunk 2]
```diff
@@ -1033,15 +1033,17 @@ static void glb_shutdown_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol) {
   while (pp != NULL) {
     pending_pick *next = pp->next;
     *pp->target = NULL;
-    GRPC_CLOSURE_SCHED(exec_ctx, &pp->wrapped_on_complete_arg.wrapper_closure,
-                       GRPC_ERROR_NONE);
+    GRPC_CLOSURE_SCHED(
+        exec_ctx, &pp->wrapped_on_complete_arg.wrapper_closure,
+        GRPC_ERROR_CREATE_FROM_STATIC_STRING("Channel Shutdown"));
     pp = next;
```
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Good catch!  Fixed.

File: src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc (hunk 1 above)
Comment (markdroth) [reply src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc]: Good catch!  Fixed.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +240 [None]                                                                                  +680  +0.0%
  +1.1%    +160 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc                          +160  +1.1%
       +36%    +155 glb_shutdown_locked                                                                     +155   +36%
      +1.4%      +5 [Unmapped]                                                                                +5  +1.4%
  +2.1%    +128 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc                  +128  +2.1%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, pick_first_lb_policy*) [clone .isra.    +192  [NEW]
  +0.4%     +32 src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc                 +32  +0.4%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, round_robin_lb_policy*) [clone .isra    +192  [NEW]
      +6.8%      +9 [Unmapped]                                                                                +9  +6.8%

  +0.0%    +560 TOTAL                                                                                  +1000  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.0%    +240 [None]                                                                                  +672  +0.0%
  +1.1%    +160 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc                          +160  +1.1%
       +36%    +155 glb_shutdown_locked                                                                     +155   +36%
      +1.4%      +5 [Unmapped]                                                                                +5  +1.4%
  +2.1%    +128 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc                  +128  +2.1%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, pick_first_lb_policy*) [clone .isra.    +192  [NEW]
  +0.4%     +32 src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.cc                 +32  +0.4%
      [NEW]    +192 fail_pending_picks_for_shutdown(grpc_exec_ctx*, round_robin_lb_policy*) [clone .isra    +192  [NEW]
      +6.8%      +9 [Unmapped]                                                                                +9  +6.8%

  +0.0%    +560 TOTAL                                                                                   +992  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
Known failures: #11055 #12510 #12957 #12917 #12960 #12931 #12961 #13085 #12932 #12886 #12918 #12964 #11274 #12963

**Pull request merged**